### ISSUE TITLE
Difficult for users to figure out how to specify dependency locations

### ISSUE THREAD

**Issue by eyalroz** [mesonbuild/meson#8581]
**Describe the bug**
When a user is trying to build a repository using meson, starting with something `meson setup build_dir`, and meson reports certain missing dependencies - the user may need to download and build those dependencies if they are unavailable as part of their OS distribution. However, once that is done, the user needs to be able to tell meson "take package `foo` from `/path/to/foo`".

Now,

* The meson binary does not tell the user how to specify a custom package location (for foo or a dependency in general)
* `meson help` and `meson help setup` do not say how to specify a custom package location.
* The `man meson` page does not say how to specify a custom package location.
* The meson reference manual does not say how to specify a custom package location - specifically, the FAQ or the "How do I do X with meson" sections don't tell you that.

**Expected behavior**

Users, who are not themselves fluet with meson and just want to build stuff which uses meson, should be able to easily access this information, which is necessary in a very common scenario.

**system parameters**
My own system is irrelevant, but just for the hell of it...
* Plain native build.
* Devuan GNU/Linux Beowulf
* python 2.7.16 and python3 3.7.3
* meson 0.49.2
* ninja 1.8.2

**Comment by eli-schwartz** [comment]
Meson just uses pkg-config. Nothing here is meson specific.

So, extraordinarily unlike cmake, where the mechanism for specifying where to find a dependency is highly cmake specific and therefore cmake needs to tell you how to do its secret special sauce.

**Comment by eli-schwartz** [comment]
And, in general, this is not information suitable for a FAQ. Instead, it has its own page at https://mesonbuild.com/Dependencies.html (which however does not tell you how pkg-config works, but does point out everything meson-specific about meson's attempt to fall back on parsing the trace output of checking for the dependency using a dummy cmake project).

**Comment by eyalroz** [comment]
> Meson just uses pkg-config. Nothing here is meson specific.

That's perfectly ok, but users should still be told this, or directed to somewhere which tells them about using `PKG_CONFIG_PATH`.

> And, in general, this is not information suitable for a FAQ. Instead, it has its own page at https://mesonbuild.com/Dependencies.html 

That page seems to be targeted at meson config file authors, telling them how to define their dependencies. If one reads down several paragraphs and examples, one does find the sentence: "The dependency detector works with all libraries that provide a pkg-config file."  - but it is easy to miss, and the document goes on to discuss specific packages.

So, I ask that at least two or three of the "venues" I mentioned above be "bolstered" with this information. Make it more idiot-proof for us idiots :-)

**Comment by eli-schwartz** [comment]
Documentation patches are welcome. :)

And `meson setup builddir` error output can be discussed (though I'm not sure how much text we should be dumping to the console about this, so I'd prefer a better idea of just what might be being proposed).

**Comment by eyalroz** [comment]
I'm not sure the user who need help figuring out PKG_CONFIG_PATH is the best person offer docs patches...

Anyway, about the error output: I'd say that adding an extra line, even a somewhat longer one, is well within reason.

Some ideas:

> If the missing package is installed in a non-standard location, and has a pkg_config (.pc) file, please add its directory to your PKG_CONFIG_PATH environment variable. 

or:

> Consult the meson documentation (https://mesonbuild.com/Dependencies.html) for information on making dependencies visible to meson.

(but what if that URL changes?)

or:

> Please invoke "meson help dependencies" for information on making dependencies visible to meson.

**Comment by eyalroz** [comment]
Oh, I should mention that adding paths to PKG_CONFIG_PATH is not enough, because one also needs to make sure that associated paths are visible after installation for dynamic loading, with `LD_LIBRARY_PATH`. And perhaps there are other ancillary settings to make.

**Comment by tkircher** [comment]
I was just trying to build gdk-pixbuf, which uses meson as a build system, and no matter what I do I can't get meson to ignore /usr/lib and look in /usr/local/lib. I can set PKG_CONFIG_PATH, LD_LIBRARY_PATH, LIBRARY_PATH, edit meson.config to require a newer version than is in /usr/lib, edit the pkgconfig files in /usr/lib to point to /usr/local/lib, and in the end I actually had to move the libraries out of /usr/lib during the build in order for meson to finally use the correct ones. The way cmake or even autotools works seems superior in every possible way.

**Comment by jpakkane** [comment]
Did you wipe the build dir (or cached data at least) when you tried again? When dependencies are found in one place, Meson will cache that and not look for them on subsequent invocations.

**Comment by tkircher** [comment]
Yeah, blew away the build dir every time, also tried configuration options like -Dlibdir=/usr/local/lib. Even when I made sure all the .pc files had no mention of /usr/lib, meson seemed to find it first. :D

**Comment by hikboy** [comment]
> Oh, I should mention that adding paths to PKG_CONFIG_PATH is not enough, because one also needs to make sure that associated paths are visible after installation for dynamic loading, with `LD_LIBRARY_PATH`. And perhaps there are other ancillary settings to make.

I meet the same problem as you, and I am a fresher about meson.
